### PR TITLE
[android] Show the speed limit like in iOS

### PR DESCRIPTION
**Pull request by alensiljak**
Implements #952, text-only

### CODE REVIEW

**Comment by alensiljak** [comment]
I have concerns about the whole value fitting once the values are in double- or triple-digits. Need to try it while driving and eventually adjust the font size for the display.

**Comment by pm4rcin** [comment]
> I have concerns about the whole value fitting once the numbers are in double- or triple-digits. Need to try it while driving and eventually adjust the font size for the display.

It could be easily tested on Android Emulator.

**Comment by alensiljak** [comment]
> It could be easily tested on Android Emulator.

I would, if you tell me how. Is there something like a "demo ride" feature from Magic Earth?

I can confirm that the display is there cause I've tested by modifying the if statement but in my location the value for the speed limit is -1 and the current speed is 0.0, obviously. That's not enough digits to really test. I'm not aware of better methods.

**Comment by pm4rcin** [comment]
You can import GPX/KML file (from Google Earth for example) or create a route like I did. I hope you understand from the screenshot:
1667596674-wayshot

**Comment by alensiljak** [comment]
Thanks, @pm4rcin, I did not know about that feature. I did test in the most simple way possible, by setting the string to "180 (130)" and checking it on the phone.

**Comment by alensiljak** [comment]
> I'll try to test by myself if I manage to patch and compile the thing (but can't promise it because I'm new to Android development).

Seems fairly straightforward by following the instructions. I did not experience any issues. The sub-repos need to be fetched and the `configure.sh` needs to be run in Linux environment, i.e. WSL if on Windows.

That's a really cool tip about the driving simulation. I'll have to try that. OsmAnd can create GPX routes easily.

**Comment by Jean-BaptisteC** [comment]
Can you add [android] before PR name to trigger Android workflow ;)
More information [here](https://github.com/organicmaps/organicmaps/blob/master/docs/PR_GUIDE.md)

**Comment by alensiljak** [comment]
@pm4rcin, could you tell me how to get the route (GPX) into Organic Maps, please? If I choose OM for opening the GPX file from i.e. Material Files, nothing happens.
This would be really handy in order to test the issues with speed cameras that I've noticed.

**Comment by pm4rcin** [comment]
@alensiljak I don't think GPX import is supported in any way currently. See https://github.com/organicmaps/organicmaps/issues/624

**Comment by pm4rcin** [comment]
What is this `(36)`? It's on motorway with 130pkh limit so it displays a different thing than wanted. When on 80kph limit it's `(22)` and on 50kph it's `(14)`. Something's wrong with your PR but it's close to displaying correct thing.
1667605175-wayshot

**Comment by alensiljak** [comment]
> What is this `(36)`?

Oh, right. The speed is apparently expressed in meters per second. I need to convert it to km/h.

**Comment by alensiljak** [comment]
Screenshot_20221105-010045_Brave

**Comment by endim8** [comment]
Nice! It'd be good to display the speed limit as iOS does, in the format `currentSpeed/speedLimit` (iOS PR for reference: https://github.com/organicmaps/organicmaps/pull/2963)

**Comment by biodranik** [comment]
@alensiljak you can convert GPX to KML using any online convertor.

**Comment by alensiljak** [comment]
Guys, thanks for all the tips so far! I now have a decent setup and can hopefully debug and check other issues that I experience, like speed cameras.

The conversion and formatting are now in place and I'd say the PR is functionally ready. Please review.

**Comment by alensiljak** [comment]
Oh, the code formatting sucks. The change is now really hard to identify on GitHub. I've tried throwing an Exception and then removed that code, reformatting the file to fix the spacing. Now the whole file is indicated as changed. :S

**Review by biodranik — CHANGES_REQUESTED** [review]
Why did you change the indentation? We have some rules for code formatting, please try to follow them. There are some code formatters at tools/android/formatter for example.

File: android/res/layout/layout_nav_bottom_numbers.xml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 <?xml version="1.0" encoding="utf-8"?>
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
-  xmlns:tools="http://schemas.android.com/tools"
-  android:layout_width="match_parent"
-  android:layout_height="wrap_content"
-  android:gravity="center_vertical"
-  android:orientation="horizontal"
-  tools:background="#3000FF00"
-  tools:ignore="RtlSymmetry">
+    xmlns:tools="http://schemas.android.com/tools"
```
Comment (biodranik) [inline android/res/layout/layout_nav_bottom_numbers.xml:3]: In all other files, the indent is 2 spaces

File: android/res/layout/layout_nav_bottom_numbers.xml [hunk 2]
```diff
@@ -17,8 +17,8 @@
       android:background="@drawable/speed_cams_bg"
       android:gravity="center"
       android:minWidth="@dimen/nav_numbers_side_min_width"
-      android:paddingStart="@dimen/nav_numbers_margin"
-      android:paddingEnd="@dimen/nav_numbers_margin">
+      android:paddingEnd="@dimen/nav_numbers_margin"
```
Comment (biodranik) [inline android/res/layout/layout_nav_bottom_numbers.xml:20]: These changes look unnecessary.

File: android/res/layout/layout_nav_bottom_numbers.xml [hunk 3]
```diff
@@ -51,8 +51,8 @@
       android:layout_weight="1.2"
       android:gravity="center"
       android:minWidth="@dimen/nav_numbers_side_min_width"
-      android:paddingStart="@dimen/nav_numbers_margin"
-      android:paddingEnd="@dimen/nav_numbers_margin">
+      android:paddingEnd="@dimen/nav_numbers_margin"
```
Comment (biodranik) [inline android/res/layout/layout_nav_bottom_numbers.xml:54]: Please don't commit unnecessary changes.

File: android/res/layout/layout_nav_bottom_numbers.xml [hunk 4]
```diff
@@ -116,34 +116,34 @@
...
   <!-- Distance -->
   <RelativeLayout
-    android:layout_width="0dp"
-    android:layout_height="match_parent"
-    android:layout_weight="1"
-    android:gravity="center"
-    android:minWidth="@dimen/nav_numbers_side_min_width"
-    android:paddingStart="@dimen/nav_numbers_margin"
-    android:paddingEnd="@dimen/nav_numbers_margin">
+      android:layout_width="0dp"
```
Comment (biodranik) [inline android/res/layout/layout_nav_bottom_numbers.xml:119]: 1. Unnecessary formatting
2. Unnecessary swap of paddingStart and paddingEnd

File: android/res/layout/layout_nav_bottom_numbers.xml [hunk 5]
```diff
@@ -116,34 +116,34 @@
...
-      android:layout_height="wrap_content"
-      android:layout_centerHorizontal="true"
-      android:includeFontPadding="false"
-      android:lines="1"
-      android:textAppearance="@style/MwmTextAppearance.NavMenu.Number"
-      tools:text="99999" />
+        android:id="@+id/distance_value"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_centerHorizontal="true"
```
Comment (biodranik) [inline android/res/layout/layout_nav_bottom_numbers.xml:131]: The same.

**Comment by alensiljak** [comment]
The question why, I believe I have answered in my [comment](https://github.com/organicmaps/organicmaps/pull/3817#issuecomment-1304547291) above.
I have now applied the project formatter. The XML file has changed more than expected. Something might be missing there. I see the value 4 for continuation indent.

**Review by biodranik — CHANGES_REQUESTED** [review]
I see only unnecessary formatting changes, without an actual improvement.

File: android/res/layout/layout_nav_bottom_numbers.xml (hunk 2 above)
Comment (alensiljak) [reply android/res/layout/layout_nav_bottom_numbers.xml:20]: I didn't make those changes. I was editing the file in the visual editor. For everything else you can blame Google and JetBrains. The formatting is applied using the project formatter for Android Studio you mentioned above.

File: android/src/app/organicmaps/widget/menu/NavMenu.java [hunk 6]
```diff
@@ -208,11 +210,49 @@ private void updateSpeedView(@NonNull RoutingInfo info)
...
+
+  private double convertMpsIntoUnits(double speedInMps, String units)
+  {
+    double result;
+
+    switch (units)
+    {
+    case "km/h":
+      // 1 m/s = 3600s/1000m = 18/5 km/h
+      result = speedInMps * 18 / 5;
```
Comment (pm4rcin) [inline android/src/app/organicmaps/widget/menu/NavMenu.java:243]: A bit off-topic but I'm wondering why do people often use 18/5 when converting to kph when it's just 3.6? Is it easier to remember or what?

Comment (alensiljak) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:243]: It is easier to copy/paste from a search result for "m/s to mph" ;)

Comment (biodranik) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:243]: `3600 / 1000` is way better and more obvious than 18/5

File: android/res/layout/layout_nav_bottom_numbers.xml (hunk 2 above)
Comment (biodranik) [reply android/res/layout/layout_nav_bottom_numbers.xml:20]: It doesn't matter who made these changes. Our repository doesn't need them. There are zero benefits in merging them. It pollutes the project git history. Please undo them.

We generally don't apply automatic code formatters, but stick to the defined code style.

File: android/src/app/organicmaps/widget/menu/NavMenu.java (hunk 6 above)
Comment (pm4rcin) [reply android/src/app/organicmaps/widget/menu/NavMenu.java]: @alensiljak also if you change it in one place change in every other to keep it consistent.

Comment (alensiljak) [reply android/src/app/organicmaps/widget/menu/NavMenu.java]: Seems that you guys have more time to work on this than I do. :D

File: android/res/layout/layout_nav_bottom_numbers.xml (hunk 2 above)
Comment (pm4rcin) [reply android/res/layout/layout_nav_bottom_numbers.xml:20]: I have reformatted this file locally and it doesn't add extra spaces. So it could be that he has non-default configuration for xml formatting or Android Studio is doing that when adjusting files by visual editor.

File: android/src/app/organicmaps/widget/menu/NavMenu.java (hunk 6 above)
Comment (pm4rcin) [reply android/src/app/organicmaps/widget/menu/NavMenu.java]: Fix the comment too.

File: android/res/layout/layout_nav_bottom_numbers.xml (hunk 2 above)
Comment (alensiljak) [reply android/res/layout/layout_nav_bottom_numbers.xml:20]: I understand and agree with your comments. But, unfortunately, I really don't have time at the moment to fiddle around with XML that was modified by the default tool for the job, using default or project settings as advised.

Comment (endim8) [reply android/res/layout/layout_nav_bottom_numbers.xml:20]: I started looking into showing speed limit too, try reset the XML, then remove `android:paddingStart` & `android:paddingEnd`lines from the speed number element

**Comment by pm4rcin** [comment]
You can reset your local branch and save the diff to the file and then `git apply filename`. Then commit and force push. WARNING: Force pushing will destroy your commit history but I don't think that we need more than 1 commit for such simple PR.

**Comment by alensiljak** [comment]
I'm getting 
```
error: patch failed: android/res/layout/layout_nav_bottom_numbers.xml:13
error: android/res/layout/layout_nav_bottom_numbers.xml: patch does not apply
```

**Comment by alensiljak** [comment]
The PR seems to have closed automatically when the branch was reset.

**Comment by alensiljak** [comment]
> Did you delete your fork's branch? Can you push it again into your fork?

See above.
I reset it before trying to apply the patch but the patch does not apply. Might need to redo the changes from the patch manually.

File: android/src/app/organicmaps/widget/menu/NavMenu.java [hunk 7]
```diff
@@ -208,11 +208,49 @@ private void updateSpeedView(@NonNull RoutingInfo info)
...
+    case "mph":
+      // 1 m/s = 2.2369 mph
+      result = speedInMps * 2.2369;
+      break;
+    default:
+      // should throw an exception. Use km/h as a workaround.
+      result = speedInMps * 3.6;
+    }
+    return result;
+  }
```
Comment (vng) [inline android/src/app/organicmaps/widget/menu/NavMenu.java:252]: Why not to uses _existing_ approach? Like:
Pair<String, String> speedAndUnits = StringUtils.nativeFormatSpeedAndUnits(last.getSpeed());

Comment (alensiljak) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:252]: In this case, the unit is already known to the caller and that would somewhat obfuscate the point of the function, I think.

File: android/src/app/organicmaps/widget/menu/NavMenu.java [hunk 8]
```diff
@@ -208,11 +208,49 @@ private void updateSpeedView(@NonNull RoutingInfo info)
...
 
+  private String formatSpeed(Pair<String, String> speedAndUnits, RoutingInfo info)
+  {
+    double currentSpeedMps = Double.parseDouble(speedAndUnits.first);
+    String speedIndicator = String.format("%.0f", currentSpeedMps);
+
+    if (info.speedLimit > 0.0)
+    {
+      double speedLimit = convertMpsIntoUnits(info.speedLimit, speedAndUnits.second);
+      speedIndicator += String.format("/%.0f", speedLimit);
```
Comment (jordane-quincy) [inline android/src/app/organicmaps/widget/menu/NavMenu.java:227]: [Pure suggestion as I would love to see this PR merged even without it and I wouldn't slow down the reviewing process]

iOS [pr](https://github.com/organicmaps/organicmaps/pull/2963/files#diff-689d1838cea59440095256c64d22da973383adbb1143f45a1c545d32621b1de4R160) use space around the slash:
```suggestion
      speedIndicator += String.format(" / %.0f", speedLimit);
```

Comment (alensiljak) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:227]: > iOS use space around the slash:

The initial implementation was like that. Unfortunately, it takes too much screen space, in my opinion. Hence I narrowed it down.

Comment (jordane-quincy) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:227]: Oh ! ok. Sorry I didn't see this from the history because of the force push.

Is there anything I can do to help the merge of this pr ?
Speed limit information is one of key features I really love to see in Organic Maps

File: android/res/layout/layout_nav_bottom_numbers.xml (hunk 1 above)
Comment (alensiljak) [reply android/res/layout/layout_nav_bottom_numbers.xml]: I believe this is resolved by the second attempt, applying only the changes without reformatting.

File: android/src/app/organicmaps/widget/menu/NavMenu.java (hunk 7 above)
Comment (biodranik) [reply android/src/app/organicmaps/widget/menu/NavMenu.java:252]: Is this duplication really necessary?

**Review by biodranik — COMMENTED** [review]
What does it look like in a vertical orientation on different devices, including lower-end ones (e.g. Android 5-6), with speed 110 and speed limit 120, and with a larger/largest font selected in the device's options?

**Review by biodranik — CHANGES_REQUESTED** [review]
https://github.com/organicmaps/organicmaps/issues/4189 is a blocker for this PR and should be fixed first.

**Comment by TheAdventurer64** [comment]
Looks like the issue was fixed

**Comment by biodranik** [comment]
@TheAdventurer64 no, this PR is for Android

**Comment by simonschaufi** [comment]
Just wanted to inform you that maps.me has implemented this feature already and their app is also open source. You might have a look at their implementation 😄

Screenshot_20230416-134804_MAPSME

And when you are too fast, it changes the color:

Screenshot_20230416-134940_MAPSME

**Comment by d4f5409d** [comment]
> Just wanted to inform you that maps.me has implemented this feature already and their app is also open source. You might have a look at their implementation 😄
> 
> https://github.com/mapsme/omim
> 
> Screenshot_20230416-134804_MAPSME
> 
> And when you are too fast, it changes the color:
> 
> Screenshot_20230416-134940_MAPSME
> 

They're proprietary. Organic Maps is the fork of the Maps.me repo.

**Comment by rtsisyk** [comment]
@alensiljak what is the current status here?

**Comment by alensiljak** [comment]
> @alensiljak what is the current status here?

Hi. I'm probably the wrong person to ask. I've implemented a simple display (i.e. "50/70 kmh") of the speed limit, if available.
It is a decent base for any more-advanced graphical display of the speed limit, where eventually the textual changes is reverted and a graphical indicator displayed instead.
No extensive testing on different devices or screen resolutions have been performed as I have neither the time nor other resources to do so.
Considering I don't know what the process for accepting pull requests is, and this has been sitting there for over a year, someone closer to the project might be able to answer that.

**Comment by Jean-BaptisteC** [comment]
An other PR to implement speed limit was already started here #5233

**Comment by alensiljak** [comment]
> An other PR to implement speed limit was already started here #5233

"Already"?
If my knowledge of numbers and dates is correct, that commit was created in October 2023 whereas my pull request was submitted in November 2022.

**Comment by AndrewShkrob** [comment]
I think this PR can be closed

There is a PR with UI implementation of speed limit: https://github.com/organicmaps/organicmaps/pull/7005
